### PR TITLE
Improved pattern scanner

### PR DESCRIPTION
**Pull request by Vali0004**
Previous time:
- Total time was 1.05779s
- Per sig average was 0.0199584s
Current time:
- Total time was 0.567962s
- Per sig average was 0.0107163s
Time Differences
- Total time difference was 0.489828s
- Per sig average difference was 0.0092421
Conclusion:
The difference can be quite drastic, but I did simple changes. More optimizations could've been done but that could've hindered code readability and understandability
More can be done to get it under 0.3s to 0.4s

### CODE REVIEW

**Review by SkiddyToast — DISMISSED** [review]
LGTM, thanks Vali!

**Comment by Vali0004** [comment]
Just added another small commit to remove something unneeded that could slow it down. I'll make one final last optimization and that should be it

File: BigBaseV2/src/memory/pattern.cpp [hunk 1]
```diff
@@ -3,67 +3,48 @@
...
+		case 'f': return static_cast<std::uint8_t>(0xf);
+		case 'A': return static_cast<std::uint8_t>(0xA);
+		case 'B': return static_cast<std::uint8_t>(0xB);
+		case 'C': return static_cast<std::uint8_t>(0xC);
+		case 'D': return static_cast<std::uint8_t>(0xD);
+		case 'E': return static_cast<std::uint8_t>(0xE);
+		case 'F': return static_cast<std::uint8_t>(0xF);
+		default: return std::nullopt;
+		}
+	}
```
Comment (Yimura) [inline BigBaseV2/src/memory/pattern.cpp:57]: newline

File: BigBaseV2/src/memory/range.cpp [hunk 2]
```diff
@@ -29,39 +29,81 @@ namespace memory
...
+					curIdx += shiftTable[*begin.add(curIdx + maxIdx).as<uint8_t*>()];
+					break;
+				}
+				else if (sigIdx == NULL)
+				{
+					return begin.add(curIdx);
+				}
+			}
+		}
+		return nullptr;
```
Comment (Yimura) [inline BigBaseV2/src/memory/range.cpp]: Variables do not follow coding style. (snake_case)

File: BigBaseV2/src/memory/range.cpp [hunk 3]
```diff
@@ -29,39 +29,81 @@ namespace memory
...
 
+	bool pattern_matches(std::uint8_t* target, const std::optional<std::uint8_t>* sig, std::size_t length) {
+		for (std::size_t i{}; i != length; ++i)
+		{
+			if (sig[i] && *sig[i] != target[i])
+				return false;
+		}
+
+		return true;
+	}
```
Comment (Yimura) [inline BigBaseV2/src/memory/range.cpp:107]: newline at end of function

File: BigBaseV2/src/memory/pattern.cpp [hunk 4]
```diff
@@ -3,67 +3,48 @@
...
+	pattern::pattern(std::string_view ida_sig)
+	{
+		for (std::size_t i{}; i != ida_sig.size(); ++i)
```
Comment (Yimura) [inline BigBaseV2/src/memory/pattern.cpp]: Does the compiler optimise this ida_sig.size() call or does it call .size() each iteration?

File: BigBaseV2/src/memory/pattern.cpp [hunk 5]
```diff
@@ -72,14 +76,14 @@ namespace memory
...
-				m_bytes.push_back(std::nullopt);
+				m_bytes.push_back({});
 			}
 		}
 	}
 
 	pattern::pattern(const void *bytes, std::string_view mask)
 	{
-		for (std::size_t i = 0; i < mask.size(); ++i)
+		for (std::size_t i{}; i != mask.size(); ++i)
```
Comment (Yimura) [inline BigBaseV2/src/memory/pattern.cpp]: Does the compiler optimise this mask.size() call or does it call .size() each iteration?

File: BigBaseV2/src/memory/range.cpp [hunk 6]
```diff
@@ -29,39 +29,81 @@ namespace memory
...
 	std::vector<handle> range::scan_all(pattern const &sig)
 	{
 		std::vector<handle> result;
 
 		auto data = sig.m_bytes.data();
 		auto length = sig.m_bytes.size();
-		for (std::uintptr_t i = 0; i < m_size - length; ++i)
+		for (std::uintptr_t i{}; i != m_size - length; ++i)
```
Comment (Yimura) [inline BigBaseV2/src/memory/range.cpp]: Does the compiler optimise this m_size - length or does it calculate this difference each iteration?

File: BigBaseV2/src/memory/batch.cpp [hunk 7]
```diff
@@ -20,7 +20,10 @@ namespace memory
 				if (entry.m_callback)
 				{
 					std::invoke(std::move(entry.m_callback), result);
-					LOG(INFO) << "Found '" << entry.m_name << "' GTA5.exe+" << HEX_TO_UPPER(result.as<DWORD64>() - region.begin().as<DWORD64>());
+					std::stringstream file_out{}; //I hate this, but not logging it except on fail really helps on speed.
+					file_out << "Found '" << entry.m_name << "' GTA5.exe+" << HEX_TO_UPPER(result.as<DWORD64>() - region.begin().as<DWORD64>());
+					big::g_log->file_out(file_out);
+					file_out.clear();
```
Comment (Yimura) [inline BigBaseV2/src/memory/batch.cpp:26]: Looks ugly but will do for now.

File: BigBaseV2/src/logger.hpp [hunk 8]
```diff
@@ -94,7 +94,10 @@ namespace big
 			if (!m_did_console_exist && m_attach_console)
 				FreeConsole();
 		}
-
+		void file_out(std::stringstream& str)
```
Comment (Yimura) [inline BigBaseV2/src/logger.hpp:98]: newline before this

**Review by Yimura — CHANGES_REQUESTED** [review]
Thanks for improving the sig scanner. 

Still a few things that require changes before we can merge them into master.

File: BigBaseV2/src/memory/pattern.cpp (hunk 1 above)
Comment (Vali0004) [reply BigBaseV2/src/memory/pattern.cpp:57]: Fixed in newest commit

File: BigBaseV2/src/memory/range.cpp (hunk 2 above)
Comment (Vali0004) [reply BigBaseV2/src/memory/range.cpp]: Will commit soon, waiting for review to see if there was anything else.

File: BigBaseV2/src/memory/range.cpp (hunk 6 above)
Comment (Vali0004) [reply BigBaseV2/src/memory/range.cpp]: I believe it does, however to be safe, I will add it to a const var

File: BigBaseV2/src/memory/batch.cpp (hunk 7 above)
Comment (Vali0004) [reply BigBaseV2/src/memory/batch.cpp:26]: True

File: BigBaseV2/src/logger.hpp (hunk 8 above)
Comment (Vali0004) [reply BigBaseV2/src/logger.hpp:98]: Fixed in soon commit

File: BigBaseV2/src/memory/pattern.cpp (hunk 4 above)
Comment (Vali0004) [reply BigBaseV2/src/memory/pattern.cpp]: It should, let me change

File: BigBaseV2/src/memory/pattern.cpp (hunk 5 above)
Comment (Vali0004) [reply BigBaseV2/src/memory/pattern.cpp]: Should do so, commiting changes now.

Comment (Vali0004) [reply BigBaseV2/src/memory/pattern.cpp]: Committed requested changes.

File: BigBaseV2/src/memory/pattern.cpp (hunk 4 above)
Comment (Vali0004) [reply BigBaseV2/src/memory/pattern.cpp]: Committed requested changes.

File: BigBaseV2/src/memory/range.cpp (hunk 3 above)
Comment (Vali0004) [reply BigBaseV2/src/memory/range.cpp:107]: Done

**Comment by Vali0004** [comment]
Changes should've been resolved, please re-review everything and ensure they were done as asked.

**Review by Yimura — DISMISSED** [review]

File: BigBaseV2/src/memory/range.cpp [hunk 9]
```diff
@@ -55,7 +55,7 @@ namespace memory
 			shift_table[i] = maxShift;
 		}
 
-		//Fill shift table
+		//Fill shift table with sig bytes
```
Comment (Vali0004) [inline BigBaseV2/src/memory/range.cpp:58]: https://uploadi.ng/uTEG0LGC

**Review by Yimura — APPROVED** [review]

**Pull request merged**